lib/Dialect/TritonGPU/Transforms/Utility.cpp

```diff
@@ -352,6 +354,53 @@ static std::optional<Attribute> inferSrcEncoding(triton::TransOp op,
                                  triton::inversePermutation(op.getOrder()));
 }
 
+static std::optional<Attribute>
+inferReshapeOpDstEncoding(ArrayRef<int64_t> srcShape, Attribute srcEnc,
+                          ArrayRef<int64_t> dstShape, bool allowReorder) {
+  // We don't do anything smart to allow-reorder reshapes here.  They are
+  // handled in OptimizeThreadLocality.
+  if (allowReorder)
+    return std::nullopt;
+
+  Attribute dstEnc;
+  if (succeeded(
+          srcEnc.getDialect()
+              .getRegisteredInterface<triton::DialectInferLayoutInterface>()
+              ->inferReshapeOpNoReorderEncoding(
+                  srcShape, srcEnc, dstShape, dstEnc, /*loc=*/std::nullopt))) {
+    return dstEnc;
+  }
+  return std::nullopt;
+}
+
+static std::optional<Attribute> inferDstEncoding(triton::ReshapeOp op,
+                                                 Attribute encoding) {
+  auto srcTy = op.getOperand().getType().cast<RankedTensorType>();
+  auto dstTy = op.getType().cast<RankedTensorType>();
+  return inferReshapeOpDstEncoding(srcTy.getShape(), encoding, dstTy.getShape(),
+                                   op.getAllowReorder());
+}
+
+static std::optional<Attribute> inferSrcEncoding(triton::ReshapeOp op,
+                                                 Attribute encoding) {
+  // The encoding of x given the encoding of y in `reshape(x) -> y` is the same
+  // as the encoding of x given the encoding of y in `reshape(y) -> x`.
+  auto srcTy = op.getOperand().getType().cast<RankedTensorType>();
+  auto dstTy = op.getType().cast<RankedTensorType>();
+  auto ret = inferReshapeOpDstEncoding(dstTy.getShape(), encoding,
+                                       srcTy.getShape(), op.getAllowReorder());
+  if (ret.has_value()) {
+    // We have that inferDstEnc(dstShape, dstEnc, srcShape) => srcEnc.  Check
+    // that it works in the opposite direction, which is what this function is
+    // nominally doing: inferDstEnc(srcShape, srcEnc, dstShape) => dstEnc.
+    auto inferredDstEnc = inferReshapeOpDstEncoding(
```

Comment:
Should this go in LLVM_DEBUG? Or better would be to test that the reversal is always true in your new unit test so that we don't have to have this code at all?

Reply:
It is checked in the unit test and the fuzzer.  OTOH I missed this property earlier, and this check caught a bug.

Am I being too paranoid?

I don't think we want it under LLVM_DEBUG because then the check will only run when you do `-debug`, which is for printing.  But maybe we want it under `#ifndef NDEBUG`?  Do we even use that macro in our build system?  I don't see it in cmake (despite having added a few of these in the code).

Reply:
yeah seems a bit redundant. Wouldn't the verifier catch it if we generate invalid combinations?

Reply:
sgtm, will remove